File: app/src/main/java/com/microsoft/msftbreweryworkshop/service/BreweryServiceImpl.java
package com.microsoft.msftbreweryworkshop.service;

import com.microsoft.msftbreweryworkshop.api.BreweryApi;
import com.microsoft.msftbreweryworkshop.api.model.BreweryDetail;
import com.microsoft.msftbreweryworkshop.api.model.BreweryItem;
import com.microsoft.msftbreweryworkshop.api.model.DetailResponse;
import com.microsoft.msftbreweryworkshop.api.model.ListResponse;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class BreweryServiceImpl implements BreweryService {
    private BreweryApi breweryApi;

    public BreweryServiceImpl(BreweryApi breweryApi) {
        this.breweryApi = breweryApi;
    }

    @Override
    public void getBreweries(final ServiceListener<ListResponse<BreweryItem>> listener) {
        breweryApi.getBreweries().enqueue(new Callback<ListResponse<BreweryItem>>() {
            @Override
            public void onResponse(Call<ListResponse<BreweryItem>> call, Response<ListResponse<BreweryItem>> response) {
                if (response.isSuccessful()) {
                    listener.onSuccess(response.body());
                } else {
                    listener.onFailure(response.message());
                }
            }

            @Override
            public void onFailure(Call<ListResponse<BreweryItem>> call, Throwable t) {
                listener.onFailure(t.getLocalizedMessage());
            }
        });
    }

    @Override
    public void getBrewery(String id, final ServiceListener<DetailResponse<BreweryDetail>> listener) {
        breweryApi.getBrewery(id).enqueue(new Callback<DetailResponse<BreweryDetail>>() {
            @Override
            public void onResponse(Call<DetailResponse<BreweryDetail>> call, Response<DetailResponse<BreweryDetail>> response) {
                if (response.isSuccessful()) {
                    listener.onSuccess(response.body());
                } else {
                    listener.onFailure(response.message());
                }
            }

            @Override
            public void onFailure(Call<DetailResponse<BreweryDetail>> call, Throwable t) {
                listener.onFailure(t.getLocalizedMessage());
            }
        });
    }
}

File: app/src/main/java/com/microsoft/msftbreweryworkshop/service/BreweryServiceImpl.kt
package com.microsoft.msftbreweryworkshop.service

import com.microsoft.msftbreweryworkshop.api.BreweryApi
import com.microsoft.msftbreweryworkshop.api.model.BreweryDetail
import com.microsoft.msftbreweryworkshop.api.model.BreweryItem
import com.microsoft.msftbreweryworkshop.api.model.DetailResponse
import com.microsoft.msftbreweryworkshop.api.model.ListResponse
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class BreweryServiceImpl(private val breweryApi: BreweryApi) : BreweryService {
    override fun getBreweries(listener: ServiceListener<ListResponse<BreweryItem>>) {
        breweryApi.getBreweries().enqueue(object : Callback<ListResponse<BreweryItem>> {
            override fun onFailure(call: Call<ListResponse<BreweryItem>>, t: Throwable) {
                listener.onFailure(t.localizedMessage)
            }

            override fun onResponse(
                call: Call<ListResponse<BreweryItem>>,
                response: Response<ListResponse<BreweryItem>>
            ) {
                if (response.isSuccessful) {
                    response.body()?.let {
                        listener.onSuccess(it)
                    }
                } else {
                    listener.onFailure(response.message())
                }
            }

        })
    }

    override fun getBrewery(id: String, listener: ServiceListener<DetailResponse<BreweryDetail>>) {
        breweryApi.getBrewery(id).enqueue(object : Callback<DetailResponse<BreweryDetail>> {
            override fun onFailure(call: Call<DetailResponse<BreweryDetail>>, t: Throwable) {
                listener.onFailure(t.localizedMessage)
            }

            override fun onResponse(
                call: Call<DetailResponse<BreweryDetail>>,
                response: Response<DetailResponse<BreweryDetail>>
            ) {
                if (response.isSuccessful) {
                    response.body()?.let {
                        listener.onSuccess(it)
                    }
                } else {
                    listener.onFailure(response.message())
                }
            }

        })
    }
}